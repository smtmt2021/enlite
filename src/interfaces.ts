/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { EnliteSocket } from './enlite-impl';
import { Controller } from './controller';
import { Device } from './device';
import { Property } from './property';
import { ESV } from './esv';
import { CLASS } from './device-class';
import { WsunConfig, WsunId, WsunCache, NameOfClass } from '@smtmt2021/wsun-adaptor';

export interface FrameParams {
   tid?: number;
   seoj?: number;
   deoj?: number;
   esv: ESV;
   properties: Property[];
 }

export interface EnliteConfig {
  wsun: {
    adaptor: NameOfClass;
    id: WsunId;
    cache: WsunCache;
    config: WsunConfig;
  }
}

export interface SendToRemote {
  send(params: FrameParams, address: string): Promise<void>;
}

export interface TimeValue {
  time: Date;
  value: number;
}

export interface CreateDevice {
  createDevice(controller: Controller, deviceClass: CLASS, address: string, instanceCode?: number): Device | undefined;
}

export interface CreateSocket {
  createSocket(): EnliteSocket;
}

export interface DeviceConstructor {
  new(controller: Controller, address: string, deviceClass: CLASS, instanceCode?: number): Device;
}

export type EnliteLocal = CreateDevice & CreateSocket;
