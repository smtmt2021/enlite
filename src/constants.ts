/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
const isTesting = process.env.NODE_ENV === 'test';
export const TIMEOUT_GET = (isTesting ? 1 : 10) * 1000;
