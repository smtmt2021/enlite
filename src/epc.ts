/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
export enum EPC {
   INSTANCE_LIST_NOTIFICATION = 0xd5,

   COEFFICIENT = 0xd3,
   EFFECTIVE_DIGITS = 0xd7,
   UNIT = 0xe1,
   INSTANTANEOUS_ELECTRIC_ENERGY = 0xe7,
   CUMULATIVE_AMOUNTS_OF_ENERGY_NORMAL = 0xea,
   CUMULATIVE_AMOUNTS_OF_ENERGY_REVERSE = 0xeb
 }
