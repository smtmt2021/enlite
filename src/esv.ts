/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
export enum ESV {
  GET = 0x62,
  GET_RES = 0x72,
  INF = 0x73,
  INFC = 0x74,
  INFC_RES = 0x7a
}
