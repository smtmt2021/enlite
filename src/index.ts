/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { Enlite } from './enlite';
import { EnliteImpl } from './enlite-impl';
export { Enlite };
export { SendToRemote as Controller } from './interfaces';
export { Device } from './device';
export { CLASS } from './device-class';
export { SmartElectricEnergyMeter } from './smart-electric-energy-meter';

export function EnliteGetInstance(): Enlite {
  return EnliteImpl.getInstance();
}
