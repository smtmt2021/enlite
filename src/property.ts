/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { EPC } from './epc';
import { num2hex } from './lib';

export class Property {
  static confirmed(property: Property): Property {
    return new Property(property.epc);
  }

  constructor(public epc: EPC, public edt?: Buffer) {}

  get pdc(): number {
    return this.edt ? this.edt.length : 0;
  }

  toString(): string {
    let result = 'epc:' + num2hex(this.epc, 1);
    result += ' pdc:' + num2hex(this.pdc, 1);
    if (this.edt) {
      result += ' edt:';
      for (const edt of this.edt) {
        result += num2hex(edt, 1) + ' ';
      }
    } else {
      result += ' ';
    }
    return result;
  }
}
