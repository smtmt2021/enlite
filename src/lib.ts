/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { FrameParams, TimeValue } from './interfaces';
import { ESV } from './esv';

/**
 *  Convert number to hex-decimal string.
 *  @param n     to be converted.
 *  @param bytes number of bytes, e.g. 2 bytes means 4 digits.
 */
export function num2hex(n: number, bytes = 2): string {
  if (n > Number.MAX_SAFE_INTEGER) {
    throw new Error(`the specified number is too large, ${n}`);
  }
  if (bytes < 1 || bytes > 8) {
    throw new TypeError(`invalid size of bytes, $${bytes}`);
  }
  return n
    .toString(16)
    .toUpperCase()
    .padStart(bytes * 2, '0');
}

export function dumpFrameParams(params: FrameParams): string {
  let result = '';
  if (params.tid) {
    result += `tid:${num2hex(params.tid, 2)} `;
  }
  if (params.seoj) {
    result += `seoj:${num2hex(params.seoj, 3)} `;
  }
  if (params.deoj) {
    result += `deoj:${num2hex(params.deoj, 3)} `;
  }
  for (const property of params.properties) {
    result += property.toString();
  }
  return result;
}

export function getResponseService(esv: ESV): ESV {
  switch (esv) {
    case ESV.GET:
      return ESV.GET_RES;
    case ESV.INFC:
      return ESV.INFC_RES;
  }
  return esv;
}

export function edt2DateValue(edt?: Buffer): TimeValue | undefined {
  if (!edt || edt.length !== 11) {
    return;
  }
  const year = edt.readUInt16BE(0);
  const month = edt.readUInt8(2) - 1;
  const day = edt.readUInt8(3);
  const hour = edt.readUInt8(4);
  const minuite = edt.readUInt8(5);
  const second = edt.readUInt8(6);
  const value = edt.readUInt32BE(7);
  const time = new Date(year, month, day, hour, minuite, second, 0);
  return { time, value };
}

export function dateValue2Buffer(time: Date, value: number): Buffer {
  const buffer = Buffer.alloc(11);
  buffer.writeUInt16BE(time.getFullYear());
  buffer.writeUInt8(time.getMonth() + 1, 2);
  buffer.writeUInt8(time.getDate(), 3);
  buffer.writeUInt8(time.getHours(), 4);
  buffer.writeUInt8(time.getMinutes(), 5);
  buffer.writeUInt8(time.getSeconds(), 6);
  buffer.writeUInt32BE(value, 7);
  return buffer;
}
