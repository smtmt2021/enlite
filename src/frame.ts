/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { FrameParams } from './interfaces';
import { Property } from './property';
import { EPC } from './epc';
import { ESV } from './esv';
import { num2hex, dumpFrameParams, getResponseService } from './lib';

const EHD = 0x1081;

export class Frame {
  static create(params: FrameParams): Frame {
    if (!params.tid || !params.seoj || !params.deoj || !params.esv) {
      throw new Error(`Frame: params are not enough ${dumpFrameParams(params)}`);
    }
    let data = num2hex(EHD, 2);
    data += num2hex(params.tid, 2);
    data += num2hex(params.seoj, 3);
    data += num2hex(params.deoj, 3);
    data += num2hex(params.esv, 1);
    data += num2hex(params.properties.length, 1); // as OPC
    for (const property of params.properties) {
      data += num2hex(property.epc, 1) + num2hex(property.pdc, 1);
      if (property.edt) {
        for (const dt of property.edt.values()) {
          data += num2hex(dt, 1);
        }
      }
    }
    return new Frame(Buffer.from(data, 'hex'));
  }

  static confirmed(frame: Frame): FrameParams {
    const properties: Property[] = [];
    for (const property of frame.properties) {
      properties.push(Property.confirmed(property));
    }
    return {
      tid: frame.tid,
      seoj: frame.deoj,
      deoj: frame.seoj,
      esv: getResponseService(frame.esv),
      properties
    };
  }

  constructor(public readonly buffer: Buffer) {}

  get ehd(): number {
    return this.buffer.readUInt16BE(0);
  }

  set tid(tid: number) {
    this.buffer.writeUInt16BE(tid, 2);
  }

  get tid(): number {
    return this.buffer.readUInt16BE(2);
  }

  get seoj(): number {
    return this.buffer.readUIntBE(4, 3);
  }

  get deoj(): number {
    return this.buffer.readUIntBE(7, 3);
  }

  get esv(): ESV {
    return this.buffer.readUInt8(10) as ESV;
  }

  get opc(): number {
    return this.buffer.readUInt8(11);
  }

  get properties(): Property[] {
    const properties: Property[] = [];
    const buffer = this.buffer.subarray(12);
    let offset = 0;
    for (let i = 0; i < this.opc; ++i) {
      const epc = buffer[offset++] as EPC;
      const pdc = buffer[offset++];
      const edt = pdc ? buffer.subarray(offset, offset + pdc) : undefined;
      properties.push(new Property(epc, edt));
      offset += pdc;
    }
    return properties;
  }

  toString(): string {
    let result = 'ehd:' + num2hex(this.ehd, 2)
      + ' tid:' + num2hex(this.tid, 2)
      + ' seoj:' + num2hex(this.seoj, 3)
      + ' deoj:' + num2hex(this.deoj, 3)
      + ' esv:' + num2hex(this.esv, 1)
      + ' opc:' + num2hex(this.opc, 1)
      + ' ';
    if (this.opc) {
      for (const property of this.properties) {
        result += property.toString();
      }
    }
    return result;
  }
}
