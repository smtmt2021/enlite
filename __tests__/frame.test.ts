/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { Frame } from '../src/frame';
import { Property } from '../src/property';
import { FrameParams } from '../src/interfaces';
import { dumpFrameParams } from '../src/lib';
import { ESV } from '../src/esv';
import { EPC } from '../src/epc';

test('creating a frame #1', () => {
  const message = Buffer.from('1081006B0EF0010EF0017301D50401028801', 'hex');
  expect(new Frame(message).toString()).toMatchSnapshot();
});

test('creating a frame #2', () => {
  const message = Buffer.from('1081000002880105FF017201E70400000223', 'hex');
  const frame = new Frame(message);
  expect(frame.toString()).toMatchSnapshot();
  frame.tid = 0x1234;
  expect(frame.toString()).toMatchSnapshot();
  frame.tid = 0xfedc;
  expect(frame.toString()).toMatchSnapshot();
});

test('creating a frame #3 and confirmed it', () => {
  const message = Buffer.from('1081029F02880105FF017402EA0B07E501121200000000FBD2EB0B07E5011212000000000009', 'hex');
  const frame = new Frame(message);
  expect(frame.toString()).toMatchSnapshot();
  const params = Frame.confirmed(frame);
  expect(dumpFrameParams(params)).toMatchSnapshot();
  expect(Frame.create(params).toString()).toMatchSnapshot();
});

test('creating a frame from params #1', () => {
  const params: FrameParams = {
    esv: ESV.GET,
    properties: [new Property(EPC.INSTANCE_LIST_NOTIFICATION, Buffer.from('01028801', 'hex'))]
  };
  expect(() => Frame.create(params)).toThrowErrorMatchingSnapshot();
  params.tid = 0x2345;
  expect(() => Frame.create(params)).toThrowErrorMatchingSnapshot();
  params.seoj = 0x678910;
  expect(() => Frame.create(params)).toThrowErrorMatchingSnapshot();
  params.deoj = 0xabcdef;
  expect(Frame.create(params).toString()).toMatchSnapshot();
});

test('creating a frame from params #2', () => {
  const params: FrameParams = {
    tid: 0x1234,
    seoj: 0x567890,
    deoj: 0xabcdef,
    esv: ESV.GET,
    properties: []
  };
  expect(Frame.create(params).toString()).toMatchSnapshot();
});
