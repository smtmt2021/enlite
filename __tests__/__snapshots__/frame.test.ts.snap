// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creating a frame #1 1`] = `"ehd:1081 tid:006B seoj:0EF001 deoj:0EF001 esv:73 opc:01 epc:D5 pdc:04 edt:01 02 88 01 "`;

exports[`creating a frame #2 1`] = `"ehd:1081 tid:0000 seoj:028801 deoj:05FF01 esv:72 opc:01 epc:E7 pdc:04 edt:00 00 02 23 "`;

exports[`creating a frame #2 2`] = `"ehd:1081 tid:1234 seoj:028801 deoj:05FF01 esv:72 opc:01 epc:E7 pdc:04 edt:00 00 02 23 "`;

exports[`creating a frame #2 3`] = `"ehd:1081 tid:FEDC seoj:028801 deoj:05FF01 esv:72 opc:01 epc:E7 pdc:04 edt:00 00 02 23 "`;

exports[`creating a frame #3 and confirmed it 1`] = `"ehd:1081 tid:029F seoj:028801 deoj:05FF01 esv:74 opc:02 epc:EA pdc:0B edt:07 E5 01 12 12 00 00 00 00 FB D2 epc:EB pdc:0B edt:07 E5 01 12 12 00 00 00 00 00 09 "`;

exports[`creating a frame #3 and confirmed it 2`] = `"tid:029F seoj:05FF01 deoj:028801 epc:EA pdc:00 epc:EB pdc:00 "`;

exports[`creating a frame #3 and confirmed it 3`] = `"ehd:1081 tid:029F seoj:05FF01 deoj:028801 esv:7A opc:02 epc:EA pdc:00 epc:EB pdc:00 "`;

exports[`creating a frame from params #1 1`] = `"Frame: params are not enough epc:D5 pdc:04 edt:01 02 88 01 "`;

exports[`creating a frame from params #1 2`] = `"Frame: params are not enough tid:2345 epc:D5 pdc:04 edt:01 02 88 01 "`;

exports[`creating a frame from params #1 3`] = `"Frame: params are not enough tid:2345 seoj:678910 epc:D5 pdc:04 edt:01 02 88 01 "`;

exports[`creating a frame from params #1 4`] = `"ehd:1081 tid:2345 seoj:678910 deoj:ABCDEF esv:62 opc:01 epc:D5 pdc:04 edt:01 02 88 01 "`;

exports[`creating a frame from params #2 1`] = `"ehd:1081 tid:1234 seoj:567890 deoj:ABCDEF esv:62 opc:00 "`;
